doc/README.md

```diff
@@ -0,0 +1,421 @@
+
+Snapshot files start with the SHA-256 hash of the content of the file
+and use a `.snapshot` extension.
+
+## Cryptography
+
+This section is based on and thus very similar to encryption of
+[files backup](../storage/doc/design.md).
+
+### Main Key
+
+Seedvault already uses [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)
+to give users a mnemonic recovery code and for generating deterministic keys.
+The derived key has 512 bits
+and Seedvault used to use the first 256 bits as an AES key to encrypt app data.
+Unfortunately, this key's usage is limited by Android's keystore to encryption and decryption.
+Therefore, the second 256 bits are be imported into Android's keystore for use with `HMAC-SHA256`,
+so that this key can act as a main key we can deterministically derive additional keys from
+by using HKDF ([RFC5869](https://tools.ietf.org/html/rfc5869)).
+These second 256 bits must not be used for any other purpose in the future.
+We use them for a main key to avoid users having to handle yet another secret.
+
+For deriving keys, we are only using the HKDF's second 'expand' step,
+because the Android Keystore does not give us access
+to the key's byte representation (required for first 'extract' step) after importing it.
+This should be fine as the input key material is already a cryptographically strong key
+(see section 3.3 of RFC 5869 above).
+
+### Key derivation overview
+
+The original entropy comes from a BIP39 seed (12 words = 128 bit size)
+obtained from Java's `SecureRandom`.
+A PBKDF SHA512 based derivation defined in BIP39 turns this into a 512 bit seed key.
+
+The derived seed key (512 bit size) gets split into two parts:
+1. legacy app data encryption key (unused) - 256 bit - first half of seed key
+2. main key - 256 bit - second half of seed key used to derive application specific keys:
+   1. App Backup: HKDF with info "app backup repoId key"
+   2. App Backup: HKDF with info "app backup gear table key"
+   3. App Backup: HKDF with info "app backup stream key"
+   4. File Backup: HKDF with info "stream key"
+   5. File Backup: HKDF with info "Chunk ID calculation"
+
+### Stream Encryption
+
+When a stream is written to backup storage,
+it starts with a header consisting of a single byte indicating the backup format version
+(currently `0x02`) followed by the encrypted payload.
+
+All data written to backup storage will be encrypted with a fresh key
+to prevent issues with nonce/IV re-use of a single key.
+
+We derive a stream key from the main key
+by using HKDF's expand step with the UTF-8 byte representation of "app backup stream key"
+as info input.
+This stream key is then used to derive a new key for each stream.
+
+Instead of encrypting, authenticating and segmenting a cleartext stream ourselves,
+we have chosen to employ the [tink library](https://github.com/tink-crypto/tink-java) for that task.
+Since it does not allow us to work with imported or derived keys,
```

Comment:
[Information] Is there a security relevance to this detail, or any other reason to keep this in mind? Is it suboptimal in some important way to solely use its AesGcmHkdfStreaming? I'm not knowledgeable in this area, so I'm not sure why this is important.

Reply:
What we are using is not the recommended high-level mentioned in their documentation, but hooks into a lower level. It felt important to mention this, so a reviewer could check what we do here and ensure it is actually safe.